packages/gatsby/src/redux/nodes.ts

```diff
@@ -512,29 +500,20 @@ export const getNodesFromCacheByValue = (
     const values = filterCache.meta.valuesAsc as Array<FilterValue>
     // It shouldn't find the targetValue (but it might) and return the index of
     // the two value between which targetValue sits, or first/last element.
-    const point = binarySearch(values, filterValue)
-    if (!point) {
-      return undefined
-    }
-    const [pivotMin, pivotMax] = point
-    // Each pivot index must have a value and a range
-    // The returned min/max index may include the lower/upper bound, so we still
-    // have to do lte checks for both values.
-    let pivotValue = values[pivotMax]
-    if (pivotValue > filterValue) {
-      pivotValue = values[pivotMin]
-    }
-
-    // Note: the pivot value _shouldnt_ match the filter value because that
-    // means the value was actually found, but those should have been indexed
-    // so should have yielded a result in the .get() above.
-
-    const [exclPivot, inclPivot] = ranges!.get(pivotValue) as [number, number]
+    const idx = binarySearch(values, filterValue)
 
-    // Note: technically, `5 <= "5" === true` but `5` would not be cached.
-    // So we have to consider weak comparison and may have to include the pivot
-    const until = pivotValue <= filterValue ? inclPivot : exclPivot
-    return new Set(nodes!.slice(0, until))
+    if (idx >= 0) {
+	// We unexpectedly found the value in the array. Inlucude up through
+	// its index in the returned range.
+        return new Set(nodes!.slice(0, idx+1))
```

Comment:
I think I need to add a test for `[1,2,2,2,2,3]` which should fail with this appraoch. Right?

Reply:
The previous approach was definitely broken for this case. I think the suggested test would be a good one. But I think attempts to test this path might be hard at the level of filterNodes because the equality filter will probably fast-path the evaluation before we hit the binarySerach?

Reply:
If we can't reach it it shouldn't be in there ;) You can use coercion, that's the reason we have to support direct hits here in the first place.

So a node `{x: 5}` and a filter `{x:{ lte: '5'}}`, or the num/str flipped, should do the trick here.

I tend to add these filter tests in `__tests__/run-sift.js` as they feel more concise there, but that's up to you.

Reply:
I see...`filterCache.meta.values{De,A}sc`  are already unique values, so we always need to do the `ranges!` indirection. My brokenness wasn't caused by any existing tests, but I added some non-type-coercion tests that do catch it.

When I went to write the specific test with type coercion, I got different behavior between the non-sift and the sift path. In particular, sift on `[{x: 1}, {x: 2}, {x: 2}, {x: 3}]` filter `{x:{ lte: '2'}}` seems to return no nodes at all. I wrote added the tests with assertion for the two different cases. Maybe not what we want...